### ISSUE TITLE
False positive of `reportUnnecessaryComparison` on match case with `int(<float>)`

### ISSUE THREAD

**Issue by q-wertz** [microsoft/pyright#11026]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**

In the following code
```python
sr = int(2.0e6)

match sr:
    case int(2.0e6):
        a = "asdf"
    case int(2.4e6):
        a = "cvbn"
    case _:
        raise ValueError

print(a)
```
a `reportUnnecessaryComparison` error is shown in strict mode ([Playground](https://pyright-play.net/?strict=true&code=M4JwBAvGCWB2AuAKATAOgAwFMBsBKAUPgLYCG8AxgBZigBc%2BYjY5JwmMCKGOu9T-YEpDAAiVgBMAZiIZMWbDkjQAWHnwGMhUEeQBuAI1gz%2B89gH11GkCWgKAaiQA2AV0wBRECAD2IQgAcQOCQSXCA)).

As soon as it is replaced by integers directly the error is gone.
```python
sr = 2000000

match sr:
    case 2000000:
        a = "asdf"
    case 2400000:
        a = "cvbn"
    case _:
        raise ValueError

print(a)
```

Not sure if some float/int thing I don't get right but I think it also should not fail in the first example.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
sr = int(2.0e6)

match sr:
    case int(2.0e6):
        a = "asdf"
    case int(2.4e6):
        a = "cvbn"
    case _:
        raise ValueError

print(a)
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

**Comment by erictraut** [comment]
Thanks for the bug report. This problem isn't specific to class patterns. It's a deeper problem related to type narrowing for literal patterns. The runtime uses equality checks for literal patterns, so pyright needs to be more conservative with its narrowing of such patterns. For example, the literal pattern `3` can narrow the subject to `Literal[3]` only if the pre-narrowed type of the subject is known to be `int`. If it is some supertype of `int`, such as `object` or `float` or `complex`, it is not safe to `Literal[3]`. Similarly, as your example shows, an `int` subject value cannot be narrowed to `Never` if the literal pattern is a float because `int` and `float` values can be equal.

My intuition is that this change could be disruptive for users, but the "mypy_primer" run seems to indicate otherwise.

This will be addressed in the next release.

**Comment by q-wertz** [comment]
Thanks a lot for the feedback and the insights 😊

**Comment by erictraut** [comment]
This is addressed in 1.1.407
